libparsec/crates/client/src/invite/greeter.rs

```diff
@@ -422,32 +422,37 @@ struct BaseGreetInitialCtx {
 
 impl BaseGreetInitialCtx {
     async fn do_wait_peer(self) -> Result<BaseGreetInProgress1Ctx, GreetInProgressError> {
-        // Wait for the other peer
-        let mut result = self._do_wait_peer().await;
-        // If the attempt was automatically cancelled by the other peer, try again once.
-        // Previous attempts are automatically cancelled when a new start greeting attempt is made.
-        // This way, the peers can synchronize themselves more easily during the wait-peer phase,
-        // without requiring the front-end to deal with it.
-        if let Err(GreetInProgressError::GreetingAttemptCancelled {
-            origin: GreeterOrClaimer::Claimer,
-            reason: CancelledGreetingAttemptReason::AutomaticallyCancelled,
-            ..
-        }) = result
-        {
-            result = self._do_wait_peer().await
+        // Loop over wait peer attempts
+        for attempt in 0.. {
+            let result = self._do_wait_peer().await;
+            // If the attempt was automatically cancelled by the other peer, try again (at most 8 times).
+            // Previous attempts are automatically cancelled when a new start greeting attempt is made.
+            // This way, the peers can synchronize themselves more easily during the wait-peer phase,
+            // without requiring the front-end to deal with it.
+            if let Err(GreetInProgressError::GreetingAttemptCancelled {
+                origin: GreeterOrClaimer::Claimer,
+                reason: CancelledGreetingAttemptReason::AutomaticallyCancelled,
+                ..
+            }) = result
+            {
+                if attempt < WAIT_PEER_MAX_ATTEMPTS {
+                    continue;
+                }
```

Comment:
The logic seems odd, if `attempt` is `>=` to `WAIT_PEER_MAX_ATTEMPTS` we advance to the next step

Reply:
Yes, so we can return the error at the next `?`.

Reply:
0e4e548241110f76284bc5f7fb82038f05eb1d43